### PR TITLE
Added file handling in DynamicForm

### PR DESCRIPTION
**Pull request by GuidoZam**
| Q               | A
| --------------- | ---
| Bug fix?        | [ ]
| New feature?    | [x ]
| New sample?      | [ ]

#### What's in this Pull Request?

Added the possibility to choose the file to be uploaded to the document library.

### CODE REVIEW

File: src/loc/pt-pt.ts [hunk 1]
```diff
@@ -369,6 +369,9 @@ define([], () => {
   "DynamicFormNumberValueMustBeGreaterThan": "O valor deve ser maior que {0}",
   "DynamicFormNumberValueMustBeBetween": "O valor deve estar entre {0} e {1}",
   "DynamicFormNumberValueMustBeLowerThan": "O valor deve ser inferior a {0}",
+  "DynamicFormChooseFileLabel": "File",
```
Comment (aaclage) [inline src/loc/pt-pt.ts:372]: Ficheiro

File: src/loc/pt-pt.ts [hunk 2]
```diff
@@ -369,6 +369,9 @@ define([], () => {
...
+  "DynamicFormChooseFileButtonText": "Select file",
```
Comment (aaclage) [inline src/loc/pt-pt.ts:373]: Selecionar ficheiro

File: src/loc/pt-pt.ts [hunk 3]
```diff
@@ -369,6 +369,9 @@ define([], () => {
...
+  "DynamicFormRequiredFileMessage": "File is required.",
```
Comment (aaclage) [inline src/loc/pt-pt.ts:374]: Ficheiro é necessário

File: src/loc/fr-ca.ts [hunk 4]
```diff
@@ -369,6 +369,9 @@ define([], () => {
   "DynamicFormNumberValueMustBeGreaterThan": "La valeur doit être supérieure à {0}",
   "DynamicFormNumberValueMustBeBetween": "La valeur doit être comprise entre {0} et {1}",
   "DynamicFormNumberValueMustBeLowerThan": "La valeur doit être inférieure à {0}",
+  "DynamicFormChooseFileLabel": "File",
```
Comment (michaelmaillot) [inline src/loc/fr-ca.ts:372]: Fichier

File: src/loc/fr-ca.ts [hunk 5]
```diff
@@ -369,6 +369,9 @@ define([], () => {
...
+  "DynamicFormChooseFileButtonText": "Select file",
```
Comment (michaelmaillot) [inline src/loc/fr-ca.ts:373]: Sélectionnez un fichier

File: src/loc/fr-ca.ts [hunk 6]
```diff
@@ -369,6 +369,9 @@ define([], () => {
...
+  "DynamicFormRequiredFileMessage": "File is required.",
```
Comment (michaelmaillot) [inline src/loc/fr-ca.ts:374]: Un fichier est requis.

File: src/loc/fr-fr.ts [hunk 7]
```diff
@@ -369,6 +369,9 @@ define([], () => {
   "DynamicFormNumberValueMustBeGreaterThan": "La valeur doit être supérieure à {0}",
   "DynamicFormNumberValueMustBeBetween": "La valeur doit être comprise entre {0} et {1}",
   "DynamicFormNumberValueMustBeLowerThan": "La valeur doit être inférieure à {0}",
+  "DynamicFormChooseFileLabel": "File",
```
Comment (michaelmaillot) [inline src/loc/fr-fr.ts:372]: Fichier

File: src/loc/fr-fr.ts [hunk 8]
```diff
@@ -369,6 +369,9 @@ define([], () => {
...
+  "DynamicFormChooseFileButtonText": "Select file",
```
Comment (michaelmaillot) [inline src/loc/fr-fr.ts:373]: Sélectionnez un fichier

File: src/loc/fr-fr.ts [hunk 9]
```diff
@@ -369,6 +369,9 @@ define([], () => {
...
+  "DynamicFormRequiredFileMessage": "File is required.",
```
Comment (michaelmaillot) [inline src/loc/fr-fr.ts:374]: Un fichier est requis.

**Comment by michaelmaillot** [comment]
Hi @GuidoZam,

Thanks for this enhancement!

After going through the code and the new props spec, I want to be sure of what you want to cover.

If I read you PR comment, you want to provide a feature that allows to upload a file to a **document library**.

But if I have a look at the `enableFileSelection` prop definition [here](https://github.com/pnp/sp-dev-fx-controls-react/pull/1625/files#diff-51ca6beb7579356bbb03cd23b2b21ac112ceddeda25e950a9561aae5e8a52954R87), it says that you can add a file to a **list item**.

I tried with both approaches and it seems that it's not working with the list item context.

Here's what I suggest:

* If you want to only cover document libraries, it would be better to clarify it in the prop definition
    * In that case, maybe adding a couple of checks in order to display the file selection button based on the `BaseTemplate` property for example
* If you want to cover also lists, it has to be covered in the code
    * You can adapt the queries to do (add / update list item + add list attachment) from the `BaseTemplate` property also


In all the cases, can you upade the control's documentation page to reflect the enhancement? You could add a screen / video that covers this new feature if you want to 🙂

File: src/controls/dynamicForm/DynamicForm.tsx [hunk 10]
```diff
@@ -411,6 +437,66 @@ export class DynamicForm extends React.Component<
     }
   };
 
+  private addFileToLibrary = async (objects: {}) => {
```
Comment (michaelmaillot) [inline src/controls/dynamicForm/DynamicForm.tsx:440]: Please specify the type of the method

File: src/controls/dynamicForm/DynamicForm.tsx [hunk 11]
```diff
@@ -411,6 +437,66 @@ export class DynamicForm extends React.Component<
...
+    const {
+      selectedFile
+    } = this.state;
+
+    const {
+      listId,
+      listItemId,
+      contentTypeId,
+      onSubmitted,
+      onBeforeSubmit,
```
Comment (michaelmaillot) [inline src/controls/dynamicForm/DynamicForm.tsx:450]: You can remove unused props

**Comment by GuidoZam** [comment]
Thanks @michaelmaillot for the reviews and the feedbacks.
I want to cover the case of adding the file when creating a new list item inside a document library, I'll try to enhance the prop description and to also add some checks.

**Comment by AJIXuMuK** [comment]
@GuidoZam - seems like I still can enable file upload for standard lists (not doc libs)

**Comment by GuidoZam** [comment]
@AJIXuMuK I've added a check using the contentTypeId property of the dynamicForm control, that should be enough to restrict the usage only to document libraries.

**Comment by michaelmaillot** [comment]
Hi @GuidoZam, thanks for the update!

I've made some comments as a review.

Sorry to insist but could you please update the DynamicForm's documentation page in order to reflect the new props added in this PR? Feel free to insert screens / gifs or not.

**Comment by GuidoZam** [comment]
Hi @michaelmaillot  I've made some improvements, can you please check if now it's all good?

**Comment by michaelmaillot** [comment]
Hi @GuidoZam,

Sounds good to me! Can you just resolve the remaining conflict before I merge your PR please?

**Comment by GuidoZam** [comment]
Hi @michaelmaillot, I've resolved the remaining conflict.
For the update regarding the doc page I cannot see anything, am I missing something?

**Comment by michaelmaillot** [comment]
I've mentioned you in the review comment [here](https://github.com/pnp/sp-dev-fx-controls-react/pull/1625/files#diff-abe3c69e4f9c0b785e0f6bd18be58cf87abde6ea714de2e2f637ff069d4adf10R62)

**Comment by GuidoZam** [comment]
@michaelmaillot sorry but I can't see any comment there, if I click on the link you inserted it just select the row 62 of the DynamicForm.md file.

**Review by michaelmaillot — CHANGES_REQUESTED** [review]

**Comment by michaelmaillot** [comment]
> @michaelmaillot sorry but I can't see any comment there, if I click on the link you inserted it just select the row 62 of the DynamicForm.md file.

Nevermind, forgot to submit my review.........🤦‍♂️

Sorry about that.

**Comment by GuidoZam** [comment]
> > @michaelmaillot sorry but I can't see any comment there, if I click on the link you inserted it just select the row 62 of the DynamicForm.md file.
> 
> Nevermind, forgot to submit my review.........🤦‍♂️
> 
> Sorry about that.

No problem, I've resolved the change request.

**Pull request merged**